### PR TITLE
Prediction and reductions

### PR DESCRIPTION
**Pull request by vboussot**
## Summary

Eight performance changes on the prediction path and the reductions it shares with `Reduce`, folded into `perf(predictor): fold an ensemble in place, and clock the loop`. Trim's kept spans are cached per (axis, start) instead of being re-sorted per patch; `ModelComposite` hands a lone model's output on as it stands and divides an ensemble's sum in place; a lone model chunk reaches the reduction as a view rather than a `torch.stack` copy; `Mean.accumulate` adds each member at its own dtype and lets the kernel promote it; `Vote` counts each candidate's votes instead of sorting the stack (no int64 index tensor), and its `working_multiple_for` is now the planes the fold holds (`4 / cases`, the attribute being the two-member worst case); `_ensure_case_state` copies a patch's header only when a patch-level inverse will read it; a border patch under `pad_value=None` pads from the 0-d minimum on the device, with no host sync; the progress bar's status is set every ten batches with `refresh=False`, and the loop charges its phases (`fetch`, `forward`, `blend`, `finalize(stream)`, `finalize(case)`, `drain`, and the writer's own `write` with `wait(write)`) to a `SweepClock` of its own, `PREDICTION_CLOCK`, printed by rank 0 in the sweep report's shape for a loop over a second. The prediction config guide documents that line.

## Measured

| Change | Script | Before -> after |
|---|---|---|
| Trim kept spans, every patch of a grid, best of 3 | `.scratch/bench_kept_box.py` | P=1,331 (544^3 at 64^3, overlap 16): 95.6 ms -> 1.0 ms; P=18,000 (720x512x512 at 1x128x128, ov 32): 15.55 s -> 16 ms; P=79,507 (2048^3 at 64^3, overlap 16): killed at 900 s -> 46 ms |
| `ModelComposite` divide, per forward on CUDA, [1, 14, 128^3] fp16 | `.scratch/bench_composite_div.py` | 1 model: 2 allocations, 112 MiB, peak 112 MiB above resident -> 1, 56 MiB, 56 MiB; 2 models: 3 allocations, 168 MiB, peak 168 MiB -> 2, 112 MiB, 112 MiB |
| Lone chunk to the reduction, [14, 256^3] fp16, 448 MiB | `.scratch/bench_stack_lone.py` | `torch.stack([x])` 53.8 ms best of 5 (53.8 to 91.2), a new storage -> `x.unsqueeze(0)` 0.0 ms, the same storage |
| `Mean.accumulate`, 32 MiB member into a float32 total | `.scratch/bench_mean_accumulate.py` | CUDA fp16/uint8/int16/int32: 1 allocation, 64 MiB above resident -> 0 allocations, 0 MiB; CPU: a 64 MiB temporary either way, wall clock the same (fp16 23 vs 27 ms best of 7, int16 19 vs 19) |
| `Vote`, 128-row slab of 512x512 members, best of 3, peak above the members | `.scratch/bench_vote.py`, `bench_vote_after.py` | CUDA uint8 K=2: 640 MiB, 13 allocations, 34 ms -> 192 MiB, 11, 2.0 ms; K=6: 1920 MiB, 37, 63 ms -> 256 MiB, 71, 16 ms. CUDA int16 K=2: 768 MiB, 13, 36 ms -> 192 MiB, 11, 2.6 ms; K=6: 2304 MiB, 37, 70 ms -> 288 MiB, 71, 16 ms. CPU uint8 K=2: 438 -> 232 ms, K=6: 1803 -> 1190 ms; int16 K=2: 438 -> 266 ms, K=6: 1900 -> 1184 ms (load average 38 to 48 at the time) |
| `_ensure_case_state`, 18,000 thin 2.5D patches, no patch transform, 8 keys per header, best of 5 | `.scratch/bench_case_state.py` | 473 ms, 18,000 headers held -> 17 ms, 1 header held (the 17 ms is the accumulator reading its 18,000 slices) |
| Border-patch padding, 100^3 fp16 case at 32^3, 64 patches, 37 padded, `torch.cuda.set_sync_debug_mode` | `.scratch/bench_read_plan_sync.py` | 37 host syncs during the reads -> 0 |
| Progress-bar status, per call | `.scratch/bench_description.py` | `gpu_info` 46 us, `get_memory_info` 40 us, `set_description(refresh=True)` 8 us against 0 us deferred: ~95 us per batch -> ~10 us per batch |

`torch.mode` was measured and stays out of `Vote`: its tie-break differs by device, 4501 of 28672 voxels of a six-member fold on CUDA against the CPU's smallest label.

**Values:** nothing moves. The Trim boxes are the same window nonzero run, and the sha256 of every box of the first two grids is equal before and after. `torch.equal` pins the arithmetic against the previous spelling kept as the oracle: a three-model ensemble against the out-of-place quotient; the `Mean` sum for every dtype float32 promotes over (fp16, bf16, every integer width; a float64 member is still narrowed first, as before); `Vote` for two to nine members of uint8, int16, int32 and float32 labels on CPU and CUDA, against the sorted spelling; the padded patches on float32, fp16, int16 and uint8, against the `.item()` spelling. The lone chunk handed on is the assembled volume itself, and nothing else holds the buffer; a case with a patch inverse gets the same per-patch header copies of the same source, each patch undone on its own copy. The clock adds two `perf_counter` calls per phase and touches nothing in the data path.

**Tests:** `tests/unit/test_patching.py` (`test_trim_kept_boxes_are_cached_per_axis_position_not_per_patch`, `test_border_patches_pad_with_the_patch_minimum_kept_on_the_device`), `tests/unit/test_case_reduction.py` (`test_mean_promotes_each_member_inside_the_add`, `test_vote_counts_every_candidate_instead_of_sorting_the_stack`), `tests/unit/test_predictor_memory.py` (`test_model_composite_hands_over_a_lone_model_output_and_folds_an_ensemble_in_place`, `test_patch_headers_are_copied_only_for_a_patch_inverse_to_undo_on`, `test_get_output_hands_the_assembled_volume_on_as_a_view`, `test_predictor_runs_prediction_logging_once_per_batch_even_with_multiple_outputs`, `test_prediction_loop_refreshes_its_status_every_tenth_batch_and_clocks_its_phases`, `test_prediction_report_closes_on_the_loop_s_own_thread`), `tests/unit/test_predictor.py` (`test_async_writer_charges_its_writes_to_the_writer_s_own_phase`).

**Stack:** Stacked on `pr/1.8.2-04-metrics`; merge after it. Before deleting this branch, retarget the next PR of the stack.

## Summary by CodeRabbit

* **New Features**
  * Added prediction timing reports covering fetching, processing, blending, writing, queue waits, and GPU synchronization.
  * Progress descriptions now update periodically during prediction runs.

* **Performance Improvements**
  * Reduced memory use and unnecessary copying during prediction and result aggregation.
  * Improved patch blending, ensemble reduction, and vote processing efficiency.
  * Optimized asynchronous output writing and finalization.

* **Bug Fixes**
  * Preserved configured padding behavior while improving border-patch handling across data types.

* **Documentation**
  * Documented prediction timing output and the conditions under which it is displayed.

### CODE REVIEW

**Comment by vboussot** [comment]
@coderabbitai review

File: konfai/predictor.py [hunk 1]
```diff
@@ -638,7 +652,14 @@ def _ensure_case_state(
...
                 self._report_once(path, f"streaming: case '{input_dataset.name}' takes the {path} path.")
-        self.attributes[index_dataset][index_augmentation] = {}
+        # Everything past this point reads the header at index 0; a patch-level inverse reads one
+        # per patch, each undone on a copy of its own taken before any inverse ran. Only then are
+        # the copies made: on a grid of 18,000 thin 2.5D patches with nothing to undo they cost
+        # 473 ms and as many dicts held per (case, augmentation), measured.
+        attributes = self.attributes[index_dataset][index_augmentation] = {0: source_attribute}
+        if self._patch_inverses(dataset):
+            for i in range(1, len(input_dataset.patch.get_patch_slices(index_augmentation))):
+                attributes[i] = Attribute(source_attribute)
```
[the remark this answers is not in the document]
Comment (vboussot) [reply konfai/predictor.py:662]: Checked and refuted. Every patch of a (case, augmentation) carries the same case attribute object (`data_manager.py` hands `dataset.cache_attributes[a]` into each `DataItem`), and a forward patch transform runs on a throwaway copy (`get_data`: `cache_attribute = Attribute(self.cache_attributes[a])`), so there is no per-patch state to preserve. The one case that would need one, a `GLOBAL_STAT` patch transform nothing case-level captures, is refused at config time by `_check_patch_transform_invertible`. Pinned by test_patch_headers_are_copied_only_for_a_patch_inverse_to_undo_on.

File: tests/unit/test_patching.py [hunk 2]
```diff
@@ -638,6 +638,34 @@ def test_trim_kept_boxes_partition_the_volume(shape, patch, overlap):
...
+        for position, start in enumerate(axis_starts):
+            ones = combine.window(dim, position, len(axis_starts)).nonzero().flatten()
+            expected[dim, start] = slice(int(ones[0]), int(ones[-1]) + 1)
+    boxes = [accumulator._kept_box(patch_slice) for patch_slice in accumulator.patch_slices]
+
+    assert boxes == [tuple(expected[dim, s.start] for dim, s in enumerate(p)) for p in accumulator.patch_slices]
+    assert len(accumulator._kept) == sum(len(axis_starts) for axis_starts in starts) < len(slices)
+    assert [list(positions) for positions in accumulator._positions()] == starts
+
+
```
[the remark this answers is not in the document]
Comment (vboussot) [reply tests/unit/test_patching.py:668]: Not needed: `tests/conftest.py::_konfai_env` is autouse and already sets both variables with `monkeypatch.setenv` for every test in the suite.
